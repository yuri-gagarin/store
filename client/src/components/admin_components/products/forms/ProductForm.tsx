import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  name: string;
  description: string;
  details: string;
  price: string;
}

interface Props {
  name: string;
  price: string;
  description: string;
  details: string;
  newForm: boolean;
  handleCreateProduct(data: FormState): void;
  handleUpdateProduct(data: FormState): void;
}
const ProductForm: React.FC<Props> = ({ name, price, description, details, newForm, handleCreateProduct, handleUpdateProduct }): JSX.Element => {
  // local form state //
  const [ formState, setFormState ] = useState<FormState>({ name, price, description, details });
  // form ref //
  const productFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  // form state changes listeners //
  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      name: e.target.value
    });
  };
  const handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      price: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handleDetailsChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      details: e.target.value
    });
  };
  // submit action //
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateProduct(formState);
    } else {
      handleUpdateProduct(formState);
    }
  };
  // lifecycle hooks //
  useEffect(() => {
    if (productFormRef.current) {
      const elem = productFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [productFormRef]);
  
  return (
    <div className="productFormDiv" ref={productFormRef}>
      <Form id="productForm">
        <Form.Field>
          <label>Product name</label>
          <input 
            id="productFormNameInput"
            onChange={handleNameChange} 
            placeholder="Product name here ..." 
            value={formState.name}
          />
        </Form.Field>
        <Form.Field>
          <label>Product price</label>
          <input 
            id="productFormPriceInput"
            onChange={handlePriceChange} 
            placeholder="Product price here..." 
            value={formState.price}
          />
        </Form.Field>
        <Form.Field
          id='productFormDescInput'
          control={TextArea}
          label='Product Description'
          onChange={hadnleDescriptionChange}
          placeholder='Product description here...'
          value={formState.description}
         />
         <Form.Field
          id='productFormDetailsInput'
          control={TextArea}
          label='Product Details'
          onChange={handleDetailsChange}
          placeholder='Product details here...'
          value={formState.details}
         />
         {
           newForm 
            ? <Button 
                id="adminProductFormCreate" 
                type='submit' 
                onClick={handleSubmit} 
                content= "Create  New Product" 
              />
            : <Button 
              id="adminProductFormUpdate" 
              type='submit' 
              onClick={handleSubmit} 
              content="Update Product" 
            />
         }  
      </Form>
    </div>
  );
};

export default ProductForm;