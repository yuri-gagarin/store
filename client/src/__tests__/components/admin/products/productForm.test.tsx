import React from "react";
import ProductForm from "../../../../components/admin_components/products/forms/ProductForm";
import { mount, ReactWrapper } from "enzyme";
// helpers //

describe("PRoduct Form Component render tests", () => {
  describe("Empty Form render tests", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock data //
      const productData: IProductData = {
        _id: "",
        name: "",
        price: "",
        description: "",
        details: "",
        images: [],
        createdAt: ""
      }
      // mount form //
      wrapper = mount(
        <ProductForm  
          name={productData.name}
          price={productData.price}
          description={productData.description}
          newForm={true}
          details={productData.details}
          handleCreateProduct={jest.fn}
          handleUpdateProduct={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#adminProductFormNameInput'", () => {
      expect(wrapper.render().find("#adminProductFormNameInput").length).toEqual(1);
    });
    it("Should render '#adminProductFormPriceInput'", () => {
      expect(wrapper.render().find("#adminProductFormPriceInput").length).toEqual(1);
    });
    it("Should render '#adminProductFormDescInput'", () => {
      expect(wrapper.render().find("#adminProductFormDescInput").length).toEqual(1);
    });
    it("Should render '#adminProductFormDetailsInput'", () => {
      expect(wrapper.render().find("#adminProductFormDetailsInput").length).toEqual(1);
    });
    it("Should render '#adminProductFormCreate' Button", () => {
      expect(wrapper.render().find("#adminProductFormCreate").length).toEqual(1);
    });
    it("Should NOT render '#adminProductFormUpdate' Button'", () => {
      expect(wrapper.render().find("#adminProductFormUpdate").length).toEqual(0);
    });
  });
  
  describe("Form with valid data render tests", () => {
    let wrapper: ReactWrapper; let mockData: IProductData;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock product data //
      mockData = {
        _id: "1",
        name: "name",
        price: "100",
        description: "description",
        details: "details",
        images: [],
        createdAt: "now"
      };

      wrapper = mount(
        <ProductForm 
          name={mockData.name}
          price={mockData.price}
          description={mockData.description}
          newForm={false}
          details={mockData.details}
          handleCreateProduct={jest.fn}
          handleUpdateProduct={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#adminProductFormNameInput'", () => {
      expect(wrapper.render().find("#adminProductFormNameInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminProductFormNameInput'", () => {
      const input = wrapper.render().find("#adminProductFormNameInput");
      expect(input.val()).toEqual(mockData.name);
    });
    it("Should render '#adminProductFormPriceInput'", () => {
      expect(wrapper.render().find("#adminProductFormPriceInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminProductFormPriceInput'", () => {
      const input = wrapper.render().find("#adminProductFormPriceInput");
      expect(input.val()).toEqual(mockData.price);
    });
    it("Should render '#adminProductFormDescInput'", () => {
      expect(wrapper.render().find("#adminProductFormDescInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminProductFormDescInput'", () => {
      const input = wrapper.render().find("#adminProductFormDescInput");
      expect(input.val()).toEqual(mockData.description);
    });
    it("Should render '#adminProductFormDetailsInput'", () => {
      expect(wrapper.render().find("#adminProductFormDetailsInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminProductFormDetailsInput'", () => {
      const input = wrapper.render().find("#adminProductFormDetailsInput");
      expect(input.val()).toEqual(mockData.details);
    });
    it("Should render '#adminProductFormUpdate' Button'", () => {
      expect(wrapper.render().find("#adminProductFormUpdate").length).toEqual(1);
    });
    it("Should NOT render '#adminProductFormCreate' Button", () => {
      expect(wrapper.render().find("#adminProductFormCreate").length).toEqual(0);
    });
  });
})