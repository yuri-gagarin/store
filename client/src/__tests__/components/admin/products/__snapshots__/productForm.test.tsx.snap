// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PRoduct Form Component render tests Empty Form render tests Should properly mount 1`] = `
<ProductForm
  description=""
  details=""
  handleCreateProduct={[Function]}
  handleUpdateProduct={[Function]}
  name=""
  newForm={true}
  price=""
>
  <div
    className="productFormDiv"
  >
    <Form
      as="form"
      id="productForm"
    >
      <form
        className="ui form"
        id="productForm"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Product name
            </label>
            <input
              id="productFormNameInput"
              onChange={[Function]}
              placeholder="Product name here ..."
              value=""
            />
          </div>
        </FormField>
        <FormField>
          <div
            className="field"
          >
            <label>
              Product price
            </label>
            <input
              id="productFormPriceInput"
              onChange={[Function]}
              placeholder="Product price here..."
              value=""
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="productFormDescInput"
          label="Product Description"
          onChange={[Function]}
          placeholder="Product description here..."
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="productFormDescInput"
            >
              Product Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="productFormDescInput"
              onChange={[Function]}
              placeholder="Product description here..."
              rows={3}
              value=""
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="productFormDescInput"
                      placeholder="Product description here..."
                      rows="3"
                    />,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="productFormDescInput"
                        placeholder="Product description here..."
                        rows="3"
                      />,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="productFormDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Product description here..."
                    rows={3}
                    value=""
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="productFormDetailsInput"
          label="Product Details"
          onChange={[Function]}
          placeholder="Product details here..."
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="productFormDetailsInput"
            >
              Product Details
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="productFormDetailsInput"
              onChange={[Function]}
              placeholder="Product details here..."
              rows={3}
              value=""
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="productFormDetailsInput"
                      placeholder="Product details here..."
                      rows="3"
                    />,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="productFormDetailsInput"
                        placeholder="Product details here..."
                        rows="3"
                      />,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="productFormDetailsInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Product details here..."
                    rows={3}
                    value=""
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Create  New Product"
          id="adminProductFormCreate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="adminProductFormCreate"
                  type="submit"
                >
                  Create  New Product
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="adminProductFormCreate"
                    type="submit"
                  >
                    Create  New Product
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="adminProductFormCreate"
                onClick={[Function]}
                type="submit"
              >
                Create  New Product
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</ProductForm>
`;

exports[`PRoduct Form Component render tests Form with valid data render tests Should properly mount 1`] = `
<ProductForm
  description="description"
  details="details"
  handleCreateProduct={[Function]}
  handleUpdateProduct={[Function]}
  name="name"
  newForm={false}
  price="100"
>
  <div
    className="productFormDiv"
  >
    <Form
      as="form"
      id="productForm"
    >
      <form
        className="ui form"
        id="productForm"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Product name
            </label>
            <input
              id="productFormNameInput"
              onChange={[Function]}
              placeholder="Product name here ..."
              value="name"
            />
          </div>
        </FormField>
        <FormField>
          <div
            className="field"
          >
            <label>
              Product price
            </label>
            <input
              id="productFormPriceInput"
              onChange={[Function]}
              placeholder="Product price here..."
              value="100"
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="productFormDescInput"
          label="Product Description"
          onChange={[Function]}
          placeholder="Product description here..."
          value="description"
        >
          <div
            className="field"
          >
            <label
              htmlFor="productFormDescInput"
            >
              Product Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="productFormDescInput"
              onChange={[Function]}
              placeholder="Product description here..."
              rows={3}
              value="description"
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="productFormDescInput"
                      placeholder="Product description here..."
                      rows="3"
                    >
                      description
                    </textarea>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="productFormDescInput"
                        placeholder="Product description here..."
                        rows="3"
                      >
                        description
                      </textarea>,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="productFormDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Product description here..."
                    rows={3}
                    value="description"
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="productFormDetailsInput"
          label="Product Details"
          onChange={[Function]}
          placeholder="Product details here..."
          value="details"
        >
          <div
            className="field"
          >
            <label
              htmlFor="productFormDetailsInput"
            >
              Product Details
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="productFormDetailsInput"
              onChange={[Function]}
              placeholder="Product details here..."
              rows={3}
              value="details"
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="productFormDetailsInput"
                      placeholder="Product details here..."
                      rows="3"
                    >
                      details
                    </textarea>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="productFormDetailsInput"
                        placeholder="Product details here..."
                        rows="3"
                      >
                        details
                      </textarea>,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="productFormDetailsInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Product details here..."
                    rows={3}
                    value="details"
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Update Product"
          id="adminProductFormUpdate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="adminProductFormUpdate"
                  type="submit"
                >
                  Update Product
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="adminProductFormUpdate"
                    type="submit"
                  >
                    Update Product
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="adminProductFormUpdate"
                onClick={[Function]}
                type="submit"
              >
                Update Product
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</ProductForm>
`;
